lib-ee/emqx_ee_connector/src/emqx_ee_connector_clickhouse.erl

```diff
@@ -0,0 +1,439 @@
+            connecting
+    end.
+
+check_database_status(Connection) ->
+    clickhouse:status(Connection).
+
+%% -------------------------------------------------------------------
+%% on_query emqx_resouce callback and related functions
+%% -------------------------------------------------------------------
+
+-spec on_query
+    (resource_id(), Request, resource_state()) -> query_result() when
+        Request :: {RequestType, Data},
+        RequestType :: send_message,
+        Data :: map();
+    (resource_id(), Request, resource_state()) -> query_result() when
+        Request :: {RequestType, SQL},
+        RequestType :: sql | query,
+        SQL :: binary().
+
+on_query(
+    ResourceID,
+    {RequestType, DataOrSQL},
+    #{poolname := PoolName} = State
+) ->
+    ?SLOG(debug, #{
+        msg => "clickhouse connector received sql query",
+        connector => ResourceID,
+        type => RequestType,
+        sql => DataOrSQL,
+        state => State
+    }),
+    %% Have we got a query or data to fit into an SQL template?
+    SimplifiedRequestType = query_type(RequestType),
+    #{templates := Templates} = State,
+    SQL = get_sql(SimplifiedRequestType, Templates, DataOrSQL),
+    ClickhouseResult = execute_sql_in_clickhouse_server(PoolName, SQL),
+    transform_and_log_clickhouse_result(ClickhouseResult, ResourceID, SQL).
+
+get_sql(send_message, #{send_message_template := PreparedSQL}, Data) ->
+    emqx_plugin_libs_rule:proc_tmpl(PreparedSQL, Data);
+get_sql(_, _, SQL) ->
+    SQL.
+
+query_type(sql) ->
+    query;
+query_type(query) ->
+    query;
+%% Data that goes to bridges use the prepared template
+query_type(send_message) ->
+    send_message.
+
+%% -------------------------------------------------------------------
+%% on_batch_query emqx_resouce callback and related functions
+%% -------------------------------------------------------------------
+
+-spec on_batch_query(resource_id(), BatchReq, resource_state()) -> query_result() when
+    BatchReq :: nonempty_list({'send_message', map()}).
+
+on_batch_query(
+    ResourceID,
+    BatchReq,
+    State
+) ->
+    %% Currently we only support batch requests with the send_message key
+    {Keys, ObjectsToInsert} = lists:unzip(BatchReq),
+    ensure_keys_are_of_type_send_message(Keys),
+    %% Pick out the SQL template
+    #{
+        templates := Templates,
+        poolname := PoolName
+    } = State,
+    %% Create batch insert SQL statement
+    SQL = objects_to_sql(ObjectsToInsert, Templates),
+    %% Do the actual query in the database
+    ResultFromClickhouse = execute_sql_in_clickhouse_server(PoolName, SQL),
+    %% Transform the result to a better format
+    transform_and_log_clickhouse_result(ResultFromClickhouse, ResourceID, SQL).
+
+ensure_keys_are_of_type_send_message(Keys) ->
+    case lists:all(fun is_send_message_atom/1, Keys) of
+        true -> ok;
+        false -> erlang:error(<<"Unexpected type for batch message (Expected send_message)">>)
```

Comment:
Maybe we could return an error value like `{error, {unrecoverable_error, Msg}}` so that the buffer workers won't retry this batch again?

Reply:
I think this is probably a good suggestion (but I don't understand the bridge interface well enough to be certain). One problem that I have noticed while doing this bridge is that the interface to the bridge is difficult to understand. For example, how would a bridge creator know this without studying the the code for buffer workers in detail? Is this documented somewhere already? I have in my backlog to create a tutorial for creating bridges so I will need to understand the interface to make a good tutorial :) The buffer worker already catches errors in the callback. Would it be a good idea to define some special error type that has the same effect as returning `{error, {unrecoverable_error, Msg}}`? I think this could make the code of some bridges cleaner.

Reply:
Fixed by changing the bridge interface to handle special types of expectation. See the code change in the buffer worker. What do you think about this change?

Reply:
True, we indeed don't have "internal tutorials", and often I also get bogged down trying to understand some flows... :see_no_evil: 

About this `unrecoverable_error` thingy, I just know about that because I spent a lot of time trying to refactor some pieces of that code (and introduced this concept with the help of Stone).